### ISSUE TITLE
floxEnv: decide whether to eval environment variables

### ISSUE THREAD

**Issue by mkenigs** [flox/floxpkgs#82]
There are two different ways we could implement environment variables for shells:
1. let the shell evaluate them. If a user sets `foo = "$bar"`, then foo gets set to the value of bar
2. don't evaluate them. If a user sets `foo = "$bar"`, then foo gets set to literal `"$bar"`

devshell supports both (https://numtide.github.io/devshell/modules_schema.html#env), but I think that's too complicated

I'm in favor of 2:
- if users need the functionality of 1, they can do it in a shell hook
- we would have to order environment variables, which seems bug prone if we allow module composition. Example: `[FOO = "bar"] ++ [FOO = "baz" ]` != `[FOO = "baz"] ++ [FOO = "bar" ]`
- this will be more compatible with the OCI runtime spec (`env` in https://github.com/opencontainers/runtime-spec/blob/main/config.md#process)

**Comment by tomberek** [comment]
Agreed... the envs are static. Dynamic stuff can go into the hooks.

**Comment by limeytexan** [comment]
I disagree. We need to present a simple interface to the users, they should be able to set env variables without writing a "shell script" in the form of a hook. In fact the vast majority of environments won't involve hooks at all (the "default" environment in particular).

I agree that devshell is overcomplicated. What is the problem/complexity with allowing variable references? Surely we're just implementing a loop to iterate through an array of variable assignments in order, in which case the values of previous variable assignments are available for future ones. How are you proposing for that implementation to be different and why is it a problem to support this functionality?

**Comment by mkenigs** [comment]
I'm guessing this would be better to discuss live, but:

> What is the problem/complexity with allowing _sequential_ variable assignments? 

Composition, as in point 2 in my original comment. And a sequential list of variable assignments is essentially just a shell script split on newlines without `export` at the start of every element - that just feels like introducing as much complexity as we're saving users (they have to learn the concept of ordered lists of environment variables, and it just saves them typing export)

> Not sure what exactly in there speaks to this question, but we should take great care not to repeat the same mistakes of the past. Docker requires you to distinguish "arg" from "env" variables because they're combining the specification of build and execution environments in a single file, but without the scoping provided by "hooks" which can easily differentiate the use of environment variables for different purposes. We have a much better way of specifying such things, so the OCI spec may have little or no relevance here.

I'm not quite sure what you're talking about with arg vs env. According to OCI and IEEE an environment variable is a string. Evaluating an environment variable only makes sense in the context of a shell, and a container does not necessarily contain a shell
